File: Python/ASP Advanced/Arrays/Prefix Sums/Leetcode-560.py
class Solution(object):
    def subarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        hashMap = {0:1}

        res = 0
        total = 0
        for i in range(len(nums)):
            total += nums[i]
            
            if (total - k) in hashMap:
                res += hashMap[total - k]

            hashMap[total] = 1 + hashMap.get(total, 0)

        return res
        

sol = Solution()
nums = [1, -1, 0]
k = 0
print(sol.subarraySum(nums, k))

File: ASP Advanced/Arrays/Two Pointers/Leetcode-560.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int, int> umap;

        umap[0] = 1;
        int res = 0;
        int csum = 0;
        for (int num: nums){
            csum += num;
            res += umap[csum - k];
            umap[csum]++;
        }
        return res;
    }
};
// The code defines a solution to the problem of finding the number of continuous subarrays that sum to a given value k.
// It uses a hash map to store cumulative sums and their frequencies, allowing for efficient lookup of the number of subarrays that meet the criteria.
// The function iterates through the array, updating the cumulative sum and checking how many times the required sum has been seen before. 
// The result is returned as the total count of such subarrays.